anvil/src/eth/api.rs

```diff
@@ -1931,6 +1989,279 @@ impl EthApi {
     }
 }
 
+// === impl EthApi Otterscan endpoints ===
+use super::otterscan::*;
+
+impl EthApi {
+    /// Otterscan currently requires this endpoint, even though it's not part of the ots_*
+    /// https://github.com/otterscan/otterscan/blob/071d8c55202badf01804f6f8d53ef9311d4a9e47/src/useProvider.ts#L71
+    ///
+    /// It exists only as a faster alternative to eth_getBlockByNumber (by excluding uncle block
+    /// information), which is not relevant in the context of an anvil node
+    pub async fn erigon_get_header_by_number(
+        &self,
+        number: BlockNumber,
+    ) -> Result<Option<Block<TxHash>>> {
+        node_info!("ots_getApiLevel");
+
+        // TODO: probably only return blocks from before the fork?
+        self.backend.block_by_number(number).await
+    }
+
+    /// Simple API versioning scheme for the ots_* namespace
+    /// As per the latest Otterscan source code, at least version 8 is needed
+    /// https://github.com/otterscan/otterscan/blob/071d8c55202badf01804f6f8d53ef9311d4a9e47/src/params.ts#L1C2-L1C2
+    pub async fn ots_get_api_level(&self) -> Result<u64> {
+        node_info!("ots_getApiLevel");
+
+        // as required by current otterscan's source code
+        Ok(8)
+    }
+
+    /// Trace internal ETH transfers, contracts creation (CREATE/CREATE2) and self-destructs for a
+    /// certain transaction.
+    pub async fn ots_get_internal_operations(
+        &self,
+        hash: H256,
+    ) -> Result<Vec<OtsInternalOperation>> {
+        node_info!("ots_getInternalOperations");
+
+        self.backend
+            .mined_parity_trace_transaction(hash)
+            .map(OtsInternalOperation::batch_build)
+            .ok_or_else(|| BlockchainError::DataUnavailable)
+    }
+
+    /// Check if an ETH address contains code at a certain block number.
+    pub async fn ots_has_code(&self, address: Address, block_number: BlockNumber) -> Result<bool> {
+        node_info!("ots_hasCode");
+        let block_id = Some(BlockId::Number(block_number));
+        Ok(self.get_code(address, block_id).await?.len() > 0)
+    }
+
+    /// Trace a transaction and generate a trace call tree.
+    pub async fn ots_trace_transaction(&self, hash: H256) -> Result<Vec<Trace>> {
+        node_info!("ots_traceTransaction");
+        // TODO: is this right?
+        // Maybe needs to be a tree
+        self.backend.trace_transaction(hash).await
```

Comment:
I think it should be correct—it will either retrieve parity traces or recover them from the forking backend depending on usage.

Reply:
the question is more about the type of collection. currently returning an array, but should perhaps be converted to a tree
I need to dig more into the UI codebase to see what they expect here, the spec had no concrete examples

Reply:
Gotcha. I think this should be correct, just lmk what you find

Reply:
seems to be correct. the frontend code is the one converting from flat list to tree
https://github.com/otterscan/otterscan/blob/071d8c55202badf01804f6f8d53ef9311d4a9e47/src/useErigonHooks.ts#L365